File: coolero/services/sleep_listener.py
#  Coolero - monitor and control your cooling and other devices
#  Copyright (c) 2022  Guy Boldon
#  |
#  This program is free software: you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation, either version 3 of the License, or
#  (at your option) any later version.
#  |
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#  |
#  You should have received a copy of the GNU General Public License
#  along with this program.  If not, see <https://www.gnu.org/licenses/>.
# ----------------------------------------------------------------------------------------------------------------------

import logging.config
import time
from typing import Callable, List

from PySide6.QtCore import QThread
from apscheduler.job import Job
from jeepney import DBusAddress, MatchRule, message_bus
from jeepney.io.blocking import DBusConnection, open_dbus_connection, Proxy

from coolero.settings import Settings, UserSettings

_LOG = logging.getLogger(__name__)


class SleepListener(QThread):
    """
    This is a service that listens for a DBus message when the computer is put into sleep/hibernation and when it comes
    back out again. There are resets of some hardware systems when resuming and Coolero needs to re-apply
    all settings after waking up.
    This service extends and runs in its own QThread.
    """
    preparing_for_sleep_mode: bool = False
    _dbus_address_sleep: DBusAddress = DBusAddress('/org/freedesktop/login1',
                                                   bus_name='org.freedesktop.login1',
                                                   interface='org.freedesktop.login1.Manager')

    def __init__(self, device_update_jobs: List[Job]) -> None:
        super().__init__()
        self._force_apply_fun: Callable = lambda: _LOG.warning("Force Apply All Settings function not set")
        self._device_update_jobs = device_update_jobs
        self._speed_scheduler_jobs: List[Job] = []
        try:
            self._connection_system: DBusConnection = open_dbus_connection(bus='SYSTEM')
            self._match_rule = MatchRule(
                type="signal",
                interface=self._dbus_address_sleep.interface,
                member="PrepareForSleep",
                path=self._dbus_address_sleep.object_path
            )
            bus_proxy = Proxy(message_bus, self._connection_system)
            _LOG.info("System DBus connection established: %s", bus_proxy.AddMatch(self._match_rule) == ())
            self.start()
        except BaseException as ex:
            _LOG.error('Could not open DBus connection for listening', exc_info=ex)

    def set_force_apply_fun(self, force_apply_fun: Callable) -> None:
        """This is set after all devices controls have finished initialization"""
        self._force_apply_fun = force_apply_fun

    def set_speed_scheduler_jobs(self, speed_scheduler_jobs: List[Job]) -> None:
        self._speed_scheduler_jobs = speed_scheduler_jobs

    def run(self) -> None:
        with self._connection_system.filter(self._match_rule) as queue:  # if this errors out, terminate process
            while True:
                try:
                    _LOG.debug("Listening...")
                    signal_msg = self._connection_system.recv_until_filtered(queue)
                    _LOG.debug("DBus message received: %s ; %s", signal_msg.header, signal_msg.body)
                    if signal_msg.body[0]:  # returns true if entering sleep, false when waking
                        _LOG.info("System is going to sleep/hibernating, pausing jobs")
                        SleepListener.preparing_for_sleep_mode = True
                        for job in self._device_update_jobs:
                            job.pause()
                        for job in self._speed_scheduler_jobs:
                            job.pause()
                    else:
                        _LOG.info("System is resuming from sleep/hibernate, resuming jobs")
                        if delay := Settings.user.value(UserSettings.STARTUP_DELAY, defaultValue=0, type=int):
                            # use startup delay in case usb connections take longer than normal
                            time.sleep(delay)
                        else:
                            time.sleep(1.0)  # give the system at least a moment to wake up
                        self._force_apply_fun()
                        _LOG.debug("Resuming paused jobs after reinitialization")
                        for job in self._device_update_jobs:
                            job.resume()
                        for job in self._speed_scheduler_jobs:
                            job.resume()
                        time.sleep(1)  # give jobs a moment to process before waking up fully
                        SleepListener.preparing_for_sleep_mode = False
                except BaseException as ex:
                    _LOG.error("Unexpected Error", exc_info=ex)

    def shutdown(self) -> None:
        if self.isRunning():
            self._connection_system.close()
            self.terminate()
            self.wait(3000)
        _LOG.debug("Sleep DBus Listener shutdown")

File: coolercontrold/daemon/src/sleep_listener.rs
/*
 * CoolerControl - monitor and control your cooling and other devices
 * Copyright (c) 2021-2025  Guy Boldon, Eren Simsek and contributors
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

use crate::sidecar::Sidecar;
use crate::ENV_DBUS;
use log::{info, warn};
use std::env;
use std::ops::Not;
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::Arc;
use std::time::Duration;
use tokio::time::timeout;
use tokio_util::sync::CancellationToken;
use zbus::export::ordered_stream::OrderedStreamExt;
use zbus::{Connection, Proxy};

const SLEEP_DEFAULT_BUS_NAME: &str = "org.freedesktop.login1";
const SLEEP_OBJECTPATH: &str = "/org/freedesktop/login1";
const SLEEP_INTERFACE: &str = "org.freedesktop.login1.Manager";
const SIGNAL_PREPARE_FOR_SLEEP: &str = "PrepareForSleep";
// The whole DBus handshake (connect + proxy + AddMatch for PrepareForSleep) should complete in
// tens of milliseconds on a healthy system. We cap it so a wedged logind/dbus-broker can't block
// the listener from ever connecting; on timeout we simply run deaf (the flags stay false).
const DBUS_SETUP_TIMEOUT_S: u64 = 5;

/// Tracks system sleep/resume state for the main loop. The dbus connection and signal loop run on
/// the Tokio sidecar (zbus needs a Tokio reactor); the flags are shared back to the main thread via
/// `Arc<AtomicBool>` and polled each tick. If dbus is disabled or fails to connect, the flags stay
/// false and the listener is effectively deaf.
pub struct SleepListener {
    preparing_to_sleep: Arc<AtomicBool>,
    resuming: Arc<AtomicBool>,
}

impl SleepListener {
    pub fn new(run_token: CancellationToken, sidecar: &Sidecar) -> Self {
        let preparing_to_sleep = Arc::new(AtomicBool::new(false));
        let resuming = Arc::new(AtomicBool::new(false));
        if dbus_listener_enabled().not() {
            info!("DBUS sleep listener disabled.");
            return Self {
                preparing_to_sleep,
                resuming,
            };
        }
        let preparing = Arc::clone(&preparing_to_sleep);
        let waking = Arc::clone(&resuming);
        sidecar.spawn(move || run_listener(run_token, preparing, waking));
        Self {
            preparing_to_sleep,
            resuming,
        }
    }

    pub fn is_resuming(&self) -> bool {
        self.resuming.load(Ordering::Relaxed)
    }

    pub fn resuming(&self, is_resuming: bool) {
        self.resuming.store(is_resuming, Ordering::Relaxed);
    }

    pub fn is_not_preparing_to_sleep(&self) -> bool {
        self.preparing_to_sleep.load(Ordering::Relaxed).not()
    }

    pub fn preparing_to_sleep(&self, is_preparing_to_sleep: bool) {
        self.preparing_to_sleep
            .store(is_preparing_to_sleep, Ordering::Relaxed);
    }
}

fn dbus_listener_enabled() -> bool {
    env::var(ENV_DBUS)
        .ok()
        .and_then(|env_dbus| {
            env_dbus
                .parse::<u8>()
                .ok()
                .map(|bb| bb != 0)
                .or_else(|| Some(env_dbus.trim().to_lowercase() != "off"))
        })
        .unwrap_or(true)
}

/// Runs on the sidecar: connects to dbus, then sets the shared flags as `PrepareForSleep` signals
/// arrive. Returns (running deaf) if the connection cannot be established within the timeout.
async fn run_listener(
    run_token: CancellationToken,
    preparing_to_sleep: Arc<AtomicBool>,
    resuming: Arc<AtomicBool>,
) {
    // We wrap the full setup (not just the connect) because the hang has been seen on the AddMatch
    // inside `receive_signal` as well, not only on `Connection::system`.
    // See https://gitlab.com/coolercontrol/coolercontrol/-/issues/264.
    let setup = async {
        let conn = Connection::system().await?;
        let proxy = Proxy::new(
            &conn,
            SLEEP_DEFAULT_BUS_NAME,
            SLEEP_OBJECTPATH,
            SLEEP_INTERFACE,
        )
        .await?;
        let signal = proxy.receive_signal(SIGNAL_PREPARE_FOR_SLEEP).await?;
        Ok::<_, zbus::Error>((conn, signal))
    };
    let (conn, mut sleep_signal) =
        match timeout(Duration::from_secs(DBUS_SETUP_TIMEOUT_S), setup).await {
            Ok(Ok(pair)) => pair,
            Ok(Err(err)) => {
                warn!("Could not connect to DBUS, sleep listener will not work: {err}");
                return;
            }
            Err(_) => {
                warn!(
                    "DBUS sleep listener setup timed out after {DBUS_SETUP_TIMEOUT_S}s; \
                     continuing without it. Sleep/resume events will not be handled on this run."
                );
                return;
            }
        };
    info!("DBUS sleep listener connected.");
    loop {
        tokio::select! {
            () = run_token.cancelled() => break,
            Some(msg) = sleep_signal.next() => {
                // true when entering sleep, false when waking.
                match msg.body().deserialize::<bool>() {
                    Ok(true) => {
                        info!("Received message that system is going to sleep.");
                        preparing_to_sleep.store(true, Ordering::Relaxed);
                    }
                    Ok(false) => {
                        info!("Received message that system is waking from sleep");
                        resuming.store(true, Ordering::Relaxed);
                    }
                    Err(err) => warn!("Failed to read PrepareForSleep signal body: {err}"),
                }
            },
            else => break,
        }
    }
    let _ = conn.close().await;
}
